paddlenlp/trainer/ckpt_converter.py

```diff
@@ -0,0 +1,1083 @@
+# Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+import re
+from functools import reduce
+
+import paddle
+from paddle.distributed.checkpoint.load_state_dict import (
+    _load_state_dict,
+    get_rank_to_read_files,
+)
+from paddle.distributed.checkpoint.metadata import (
+    LocalTensorIndex,
+    LocalTensorMetadata,
+    Metadata,
+)
+from paddle.distributed.checkpoint.utils import flatten_state_dict
+from paddle.distributed.fleet.utils.log_util import logger
+
+MODEL_WEIGHT_SUFFIX = ".pdparams"
+OPTIMIZER_WEIGHT_SUFFIX = ".pdopt"
+SCHEDULER_NAME = "scheduler.pdparams"
+MODEL_META_FILE_NAME = "model_meta.json"
+OPTIMIZER_STATE_NAME_SUFFIX = [".moment1", ".moment2", ".beta1_pow_acc", ".beta2_pow_acc", ".master_weight"]
+
+
+class CheckpointConverter:
+    def __init__(self, hybrid_parallel_ckpt_path, state_dict, parameter_to_structured_name, patch_dict=None):
+        self.use_dist = True if paddle.distributed.get_world_size() > 1 else False
+        self.path = hybrid_parallel_ckpt_path
+        self.auto_parallel_state_dict = self.flatten_state_dict(state_dict)
+        self.parameter_to_structured_name = self.gather_global_object(parameter_to_structured_name)
+        model_state_global_shape = {}
+        for k, v in self.auto_parallel_state_dict.items():
+            model_state_global_shape[k] = v.shape
+        self.model_state_global_shape = self.gather_global_object(model_state_global_shape)
+        self.cur_rank = paddle.distributed.get_rank()
+
+        self.save_sharded_model = self.get_save_sharded_model_flag()
+
+        (
+            self.cur_rank_model_state_file_names,
+            self.cur_rank_optimizer_state_file_names,
+        ) = self.get_local_checkpoint_file_names()
+
+        self.global_model_state_file_names = self.gather_global_object(self.cur_rank_model_state_file_names)
+
+        self.global_optimizer_state_file_names = self.gather_global_object(self.cur_rank_optimizer_state_file_names)
+
+        self.initial_distributed_configuration()
+        logger.debug(
+            f"The current checkpoint’s distributed strategy is tp{self.tp_degree}, pp{self.pp_degree}, sharding{self.sharding_degree}"
+        )
+
+        self.patch_dict = patch_dict
+        for k, v in self.parameter_to_structured_name.items():
```

Comment:
Add check: '''if self.patch_dict is not None:'''

Reply:
done